integration/kotlinx-coroutines-androidx-lifecycle/src/LifecycleDefaultScopes.kt

```diff
@@ -0,0 +1,52 @@
+package kotlinx.coroutines.androidx.lifecycle
+
+import androidx.lifecycle.Lifecycle
+import androidx.lifecycle.LifecycleOwner
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.Job
+import kotlinx.coroutines.SupervisorJob
+import java.util.concurrent.ConcurrentHashMap
+
+/**
+ * Returns a [CoroutineScope] that uses [Dispatchers.Main] by default, and that is cancelled when
+ * the [Lifecycle] reaches [Lifecycle.State.DESTROYED] state.
+ *
+ * Note that this value is cached until the Lifecycle reaches the destroyed state.
+ */
+val Lifecycle.coroutineScope: CoroutineScope
+    get() = cachedLifecycleCoroutineScopes[this] ?: job.let { job ->
+        val newScope = CoroutineScope(job + Dispatchers.Main)
+        if (job.isActive) {
+            cachedLifecycleCoroutineScopes[this] = newScope
+            job.invokeOnCompletion { _ -> cachedLifecycleCoroutineScopes -= this }
+        }
+        newScope
+    }
+
+/**
+ * Calls [Lifecycle.coroutineScope] for the [Lifecycle] of this [LifecycleOwner].
+ *
+ * This is an inline property, just there for convenient usage from any [LifecycleOwner],
+ * like FragmentActivity, AppCompatActivity, Fragment and LifecycleService.
+ */
+inline val LifecycleOwner.coroutineScope get() = lifecycle.coroutineScope
+
+/**
+ * Returns a [SupervisorJob] that will be cancelled as soon as the [Lifecycle] reaches
+ * [Lifecycle.State.DESTROYED] state.
+ *
+ * Note that this value is cached until the Lifecycle reaches the destroyed state.
+ *
+ * You can use this job for custom [CoroutineScope]s, or as a parent [Job].
+ */
+val Lifecycle.job: Job
+    get() = cachedLifecycleJobs[this] ?: createJob().also {
+        if (it.isActive) {
+            cachedLifecycleJobs[this] = it
+            it.invokeOnCompletion { _ -> cachedLifecycleJobs -= this }
+        }
+    }
+
+private val cachedLifecycleJobs = ConcurrentHashMap<Lifecycle, Job>()
+private val cachedLifecycleCoroutineScopes = ConcurrentHashMap<Lifecycle, CoroutineScope>()
```

Comment:
Does it make sense to cache the scopes? Otherwise, this becomes a problem:

```
// Runs initially...
lifecycle.coroutineScope.launch {
    someFunctionThatThrowsException()
}

// Runs after the initial exception...
lifecycle.coroutineScope.launch {
    somePerfectlyFineFunction
}
```

The second task simply wouldn't run because the two share the same `Job`, which would be in a cancelled state due to the exception.

Reply:
@dlew This is a `SupervisorJob()`, so it is cancelled only if you explicitly call `cancel()` on it, unlike `Job()`.